### ISSUE TITLE
Titan/Hazelcast hangs the Gremlin REPL on exit.

### ISSUE THREAD

**Issue by okram** [thinkaurelius/titan#427]
```
gremlin> g.shutdown()
Could not commit transaction due to exception during persistence
Display stack trace? [yN]
gremlin> exit
~/software/aurelius/titan$
```

I have to CNTRL-C out. Note that when I g.shutdown() I get an exception as well. Will post that in another ticket.

**Comment by okram** [comment]
Weird. Sometimes tx g.shutdown() works -- sometimes it doesn't. I don't know why. I know that g.commit() is called in g.shutdown() and perhaps related to the "Could not commit tx error." If it does work, then exit still hangs with a long stall where it looks like memory is being freed.

```
gremlin> g.shutdown()
==>null
gremlin> exit
Oct 14, 2013 2:39:36 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=383.6M, free=66.1M, total=449.6M, max=455.1M, used/total=85.30% used/max=84.27%
cpu process-load=0.00%, system-load=4.00%, system-loadaverage=315.00%
Oct 14, 2013 2:39:46 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=383.6M, free=66.0M, total=449.6M, max=455.1M, used/total=85.31% used/max=84.28%
cpu process-load=0.00%, system-load=5.00%, system-loadaverage=322.00%
Oct 14, 2013 2:39:56 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=383.7M, free=65.9M, total=449.6M, max=455.1M, used/total=85.33% used/max=84.30%
cpu process-load=0.00%, system-load=7.00%, system-loadaverage=311.00%
Oct 14, 2013 2:40:06 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=383.8M, free=65.9M, total=449.6M, max=455.1M, used/total=85.35% used/max=84.32%
cpu process-load=0.00%, system-load=5.00%, system-loadaverage=340.00%
Oct 14, 2013 2:40:16 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=383.8M, free=65.8M, total=449.6M, max=455.1M, used/total=85.36% used/max=84.33%
cpu process-load=0.00%, system-load=6.00%, system-loadaverage=327.00%
Oct 14, 2013 2:40:26 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=383.9M, free=65.7M, total=449.6M, max=455.1M, used/total=85.38% used/max=84.35%
cpu process-load=0.00%, system-load=7.00%, system-loadaverage=323.00%
Oct 14, 2013 2:40:36 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=384.0M, free=65.7M, total=449.6M, max=455.1M, used/total=85.40% used/max=84.37%
cpu process-load=0.00%, system-load=4.00%, system-loadaverage=319.00%
Oct 14, 2013 2:40:46 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=384.0M, free=65.6M, total=449.6M, max=455.1M, used/total=85.41% used/max=84.37%
cpu process-load=0.00%, system-load=5.00%, system-loadaverage=309.00%
Oct 14, 2013 2:40:56 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=384.1M, free=65.5M, total=449.6M, max=455.1M, used/total=85.43% used/max=84.39%
cpu process-load=0.00%, system-load=4.00%, system-loadaverage=302.00%
Oct 14, 2013 2:41:06 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.2M, free=64.4M, total=449.6M, max=455.1M, used/total=85.67% used/max=84.64%
cpu process-load=0.00%, system-load=4.00%, system-loadaverage=295.00%
Oct 14, 2013 2:41:16 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.3M, free=64.4M, total=449.6M, max=455.1M, used/total=85.68% used/max=84.65%
cpu process-load=0.00%, system-load=8.00%, system-loadaverage=273.00%
Oct 14, 2013 2:41:26 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.3M, free=64.3M, total=449.6M, max=455.1M, used/total=85.70% used/max=84.67%
cpu process-load=0.00%, system-load=5.00%, system-loadaverage=246.00%
Oct 14, 2013 2:41:36 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.4M, free=64.2M, total=449.6M, max=455.1M, used/total=85.72% used/max=84.69%
cpu process-load=0.00%, system-load=4.00%, system-loadaverage=239.00%
Oct 14, 2013 2:41:46 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.5M, free=64.2M, total=449.6M, max=455.1M, used/total=85.73% used/max=84.69%
cpu process-load=0.00%, system-load=4.00%, system-loadaverage=233.00%
Oct 14, 2013 2:41:56 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.5M, free=64.1M, total=449.6M, max=455.1M, used/total=85.75% used/max=84.71%
cpu process-load=0.00%, system-load=4.00%, system-loadaverage=236.00%
Oct 14, 2013 2:42:06 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.6M, free=64.0M, total=449.6M, max=455.1M, used/total=85.77% used/max=84.73%
cpu process-load=0.00%, system-load=7.00%, system-loadaverage=239.00%
Oct 14, 2013 2:42:16 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.7M, free=64.0M, total=449.6M, max=455.1M, used/total=85.78% used/max=84.74%
cpu process-load=0.00%, system-load=3.00%, system-loadaverage=217.00%
Oct 14, 2013 2:42:26 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.8M, free=63.9M, total=449.6M, max=455.1M, used/total=85.80% used/max=84.76%
cpu process-load=0.00%, system-load=5.00%, system-loadaverage=208.00%
Oct 14, 2013 2:42:36 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.8M, free=63.8M, total=449.6M, max=455.1M, used/total=85.81% used/max=84.78%
cpu process-load=0.00%, system-load=5.00%, system-loadaverage=192.00%
Oct 14, 2013 2:42:46 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.9M, free=63.7M, total=449.6M, max=455.1M, used/total=85.82% used/max=84.79%
cpu process-load=0.00%, system-load=2.00%, system-loadaverage=186.00%
Oct 14, 2013 2:42:56 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=385.9M, free=63.7M, total=449.6M, max=455.1M, used/total=85.83% used/max=84.79%
cpu process-load=0.00%, system-load=2.00%, system-loadaverage=157.00%
Oct 14, 2013 2:43:06 PM com.hazelcast.util.HealthMonitor
INFO: [192.168.1.124]:5701 [dev]
memory used=386.0M, free=63.7M, total=449.6M, max=455.1M, used/total=85.84% used/max=84.80%
cpu process-load=0.00%, system-load=3.00%, system-loadaverage=148.00%
... STILL GOING... :/
```

**Comment by mbroecheler** [comment]
@okram Are you still experiencing this issue? Is anybody else seeing this? Would be awesome if we could somehow reproduce this for debugging

**Comment by okram** [comment]
Yes. This still happens (at least I never saw it working). It is as if g.shutdown() doesn't do anything.

**Comment by dkuppitz** [comment]
This is not a problem of Titan/Hazelcast, I have same the problem with embedded Cassandra and now also in my Java app.

Here's my main method:

``` Java
    public static void main(final String[] args) throws Exception
    {
        try (final GraphLoader loader = new TitanLoader(args)) {
            loader.load();
        }

        System.exit(0);
    }
```

The last line in `load` is:

``` Java
logger.log(Level.INFO, "LOAD TIME :: {0} ms", (System.currentTimeMillis() - startTime));
```

I see this message in my console, so I know it finished.The overridden `close` method is:

``` Java
    @Override
    public void close() {
        this.graph.shutdown();
    }
```

I'm quite sure, that `System.exit(0)` is not the problem, thus it must be the `shutdown()` method.

**Comment by dalaro** [comment]
Fixed in #449
